# Performance computing metrics for finance

Often times, in financial  analytics, complex  mathematical  ex‐pressions need to be evaluated on  large  arrays  of  numbers.

In this kernel, we'll time the different approached used to increase performance using the magic %timeit function.

In [1]:
loops = 25000000
from math import *
a = range(1, loops)
def f(x):
    return 5 * log(x) + cos(x)**2
%timeit r = [f(x) for x in a]

17.3 s ± 344 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


The generic approach takes 17.3 s ± 344 ms per loop.

In [2]:
import numpy as np
a = np.arange(1, loops)
%timeit r = 5 * np.log(a) + np.cos(a)**2

981 ms ± 13.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


The numpy approach takes only 981 ms ± 13.4 ms per loop

In [3]:
import numexpr as ne
ne.set_num_threads(1)
f = '5 * log(a) + cos(a)**2'
%timeit r = ne.evaluate(f)

451 ms ± 11.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


The Numerical Expression (numexpr) approach takes 451 ms ± 11.2 ms per loop

In [4]:
ne.set_num_threads(4)
%timeit r = ne.evaluate(f)

168 ms ± 4.48 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


By using the built-in capabilities of numexpr to parallelize the execution of the re‐spective operation. This allows us to use all available threads of a CPU. This approach takes just 168 ms ± 4.48 ms per loop.